File: Algorithms/Dynamic Programming/rod_cutting_problem.py
INT_MIN = -32767
  
# Returns the best price for a rod of length n and prices of different pieces 
def cutRod(price, n): 
    pro = [0 for x in range(n+1)] 
    pro[0] = 0
  
    for i in range(1, n+1): 
        max_pro = INT_MIN 
        for j in range(i): 
             max_pro = max(max_pro, price[j] + pro[i-j-1]) 
        pro[i] = max_pro 
  
    return pro[n] 
  
# Driver program 
if __name__ == '__main__':
    len_price = [1, 5, 8, 9, 10, 17, 17, 20] 
    size = len(len_price) 
    print("Maximum Obtainable Value is " + str(cutRod(len_price, size))) 

File: Algorithms/Dynamic Programming/Rod_Cutting_Problem.java
// Java recursive solution for Rod cutting problem
class GFG {

	/* Returns the best obtainable price for a rod of length
	n and price[] as prices of different pieces */
	static int cutRod(int price[], int index, int n)
	{
		// base case
		if (index == 0) {
			return n * price[0];
		}
		// At any index we have 2 options either
		// cut the rod of this length or not cut
		// it
		int notCut = cutRod(price, index - 1, n);
		int cut = Integer.MIN_VALUE;
		int rod_length = index + 1;

		if (rod_length <= n)
			cut = price[index]
				+ cutRod(price, index, n - rod_length);

		return Math.max(notCut, cut);
	}

	/* Driver program to test above functions */
	public static void main(String args[])
	{
		int arr[] = { 1, 5, 8, 9, 10, 17, 17, 20 };
		int size = arr.length;
		System.out.println("Maximum Obtainable Value is "
						+ cutRod(arr, size - 1, size));
	}
}


